clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -501,4 +509,60 @@ private boolean isFetchable() {
         void onAssignment(Set<TopicPartition> assignment);
     }
 
+
+    /**
+     * Wrapper to manipulate assignment while exporting the number of partitions in a thread safe manner.
+     */
+    private static class Assignment<S> {
```

Comment:
This feels a little heavy. Could we add the volatile size field to `PartitionStates` instead?